rskj-core/src/main/java/org/ethereum/datasource/TransientMap.java

```diff
@@ -0,0 +1,212 @@
+package org.ethereum.datasource;
+
+import java.io.Serializable;
+import java.util.*;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+public class TransientMap<K, V> implements Map<K, V>, Serializable {
+        private static final long serialVersionUID = -1034234728574286014L;
+        private Map<K, V> m;
+        private transient Set<K> keySet;
+        private transient Set<Entry<K, V>> entrySet;
+        private transient Collection<V> values;
+        private transient Map<K, V> added;
+        private transient Map<K, V> changed;
+        private transient Set<K> deleted;
+
+        TransientMap(Map<? extends K, ? extends V> m) {
+            if (m == null) {
+                throw new NullPointerException();
+            } else {
+                this.m = (Map<K, V>) m;
+                added = new HashMap<>();
+                changed = new HashMap<>();
+                deleted = new HashSet<>();
+
+            }
+        }
+
+        public int size() {
+
+            if (m==null) {
+                return added.size();
+            } else {
+                // todo: mix
+                return this.m.size()-deleted.size()+added.size();
+            }
+        }
+
+        public boolean isEmpty() {
+            if (m==null)
+                return added.isEmpty();
+            return this.m.isEmpty() && added.isEmpty();
+        }
+
+        public boolean containsKey(Object key) {
+            if (added.containsKey(key))
+                return true;
+            if (m==null)
+                return false;
+            if (changed.containsKey(key))
+                return true;
+            return this.m.containsKey(key) && (!deleted.contains(key));
+
+        }
+
+        public boolean containsValue(Object val) {
+            throw new UnsupportedOperationException();
+        }
+
+        public V get(Object key) {
+            if (m==null)
+                return added.get(key);
+
+            if (deleted.contains(key))
+                return null;
+            V ret = added.get(key);
+            if (ret==null)
+                ret = changed.get(key);
+            if (ret==null)
+                ret = this.m.get(key);
+            return ret;
+        }
+
+        public V put(K key, V value) {
+            if (m==null)
+                return added.put(key,value);
+
+            V ret = m.get(key);
+            if (ret!=null) {
+                if (deleted.contains(key)) {
+                    deleted.remove(key);
+                    changed.put(key,value);
+                    return null;
+                }
+                return changed.put(key,value);
```

Comment:
I think we are note returning a correct previous value when the entry is new in `changed`, since put returns `null`.

Shouldn't we be returning the value provided by `m` in that case to comply with Map interface?